### DISCUSSION TITLE
Support Markdown for docstrings

### DISCUSSION THREAD

**Discussion by madig500** [microsoft/pylance-release#4078] [Enhancement]
Hello,

We are looking to  import a custom Library with its Definitions and Descriptions into Pylance. 

What are the steps to follow to include the Library and that the new functions with descriptions are available with Pylance?    Description of the functions of the library are in a Mark Down File.

regards,

Rolando

**Comment by rchiodo** [comment]
I'm assuming the library is written in Python?

If so, then Pylance will pick up its definitions automatically. 

Descriptions can be done with doc style comments:

```python
def foo():
   '''
   This is a function
   '''
```

Either in the source or in type stubs. For more information see the documentation here: https://microsoft.github.io/pyright/#/typed-libraries?id=docstrings

**Comment by rchiodo** [comment]
If the library is not written in python but has a python interface (like a [pyd](https://docs.python.org/3/extending/building.html#building-c-and-c-extensions) file), then you can still generate type stubs for the library. Pylance would then pick up the type stubs. 

See this documentation on how to create type stubs:
https://typing.readthedocs.io/en/latest/source/stubs.html

**Comment by madig500** [comment]
> 

HI,

Yes.   Functions are in Python  .py files ,  but descriptions of each functions are in a Mark Down file,  separate from the definition.   Basically we want to know how the reference is made between both in Pylance to retrieve it.    
Is it using docStrings?  


We are receiving the following message when Hover over our custom  functions

"[FunctionName]" is not defined Pylance(ReportUndefinedVariable)


Rolando

**Comment by rchiodo** [comment]
Yes, doc strings are how documentation are added to python files. 

Either directly in the python file itself or through type stubs. 

You'd use type stubs if you didn't want to put any other type information directly into the py files.

I believe doc strings were finalized in this pep:
https://peps.python.org/pep-0257/

**Comment by rchiodo** [comment]
There is no facility to have the doc strings in a separate file, other than a type stub.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Reply by erictraut** [reply]
What is there to discuss? I think Rich already answered the question. Documentation should be included in static doc strings either within the source file or type stub.

**Comment by rchiodo** [comment]
For this item:

> We are receiving the following message when Hover over our custom functions

"[FunctionName]" is not defined Pylance(ReportUndefinedVariable)

That may be for a number of reasons. It means that pylance doesn't understand the type that is calling that function (or it's not finding the module for a global function)

We'd need to see an example to really understand why this might be happening.

**Comment by heejaechang** [comment]
similar one - https://github.com/microsoft/pylance-release/discussions/4402
